### PR TITLE
Complete snaps development guide

### LINKED ISSUE #372: Complete Snaps development guide
The snaps documentation should include a robust development guide. This could take the place of the tutorial.

### PR DESCRIPTION
**Pull request by rekmarks**
Completes the snaps development guide. Modifies other snaps documentation to match. The development guide is a living document that we'll continue to update and improve, but this PR should address all major facets of the snaps development process.

Closes #372

### CODE REVIEW

File: docs/guide/snaps-development-guide.md [hunk 1]
```diff
@@ -1,47 +1,290 @@
...
-### Accessing the Internet in a Snap
+Before beginning the development of your snap, it's important to understand the snap lifecycle.
+Just like [service workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) or AWS lambda functions, snaps are designed to wake up in response to messages / events, and shut down when they are idle.
+We say that snaps have an "ephemeral" lifecycle; here one moment, gone the next.
+In addition, if MetaMask detects that a snap becomes unresponsive, it will be shut down.
+This does not mean that you can't create long-running snaps, but it does mean that your snaps must handle being shut down, especially when they are not within the JSON-RPC request / response cycle.
+
+A snap is considered "unresponsive" if:
+
+1. It has not received a JSON-RPC request for 30 seconds.
```
Comment (ritave) [inline docs/guide/snaps-development-guide.md:175]: This only takes effect after the initial run of the snap (initiation), if the initiation never finishes the snap will run without time limits. See https://github.com/MetaMask/snaps-skunkworks/blob/ritave/onslaught/packages/snap-examples/examples/onslaught/src/index.js

**Review by ritave — APPROVED** [review]

**Comment by ritave** [comment]
I've also noticed not all rpc call are documented (at least in the current publicly available version). One example is `snap_getAppKey`

File: docs/guide/snaps-development-guide.md [hunk 2]
```diff
@@ -1,47 +1,290 @@
...
+
+Just like dapps need to request the `eth_accounts` permission in order to access the user's Ethereum accounts, so do snaps need to request access to the sensitive methods in the snaps RPC API.
+Snaps can effectively expand the MetaMask RPC API by implementing their own using `wallet.registerRpcMessageHandler()`, but in order integrate deeply with MetaMask, you need to make use of the Snaps RPC API's [restricted methods](./snaps-rpc-api.html#restricted-methods).
+Access restriction is implemented using [EIP-2255 wallet permissions](https://eips.ethereum.org/EIPS/eip-2255), and you must specify the permissions required by your snap in the manifest's `initialPermissions` field.
+You can find an example of how to do this in the [template snap's manifest](https://github.com/MetaMask/template-snap/blob/main/snap.manifest.json).
+
+#### Accessing the Internet
 
 Snaps do not get access to any sensitive APIs or features by default, and Internet access is no exception to that rule.
 To access the Internet, you must be specify the permission `endowment:network-access` in the `initialPermissions` of your `snap.manifest.json` file.
```
Comment (mapachurro) [inline docs/guide/snaps-development-guide.md:191]: Should be: "To access the internet, you must specify the permission `endowment:network-access` in the `initialPermissions` of your `snap.manifest.json` file.

(remove 'be')

File: docs/guide/snaps-development-guide.md [hunk 3]
```diff
@@ -1,47 +1,290 @@
...
+      }
+    }
+  },
+  "initialPermissions": {},
+  "manifestVersion": "0.1"
+}
+```
+
+The manifest tells MetaMask important information about your snap, most especially where it's published (via `source.location`) and how to verify the integrity of the snap source code (by attempting to reproduce the `source.shasum` value).
+For the time being, snaps can only be published to the official npm registry, and the manifest must also match the corresponding fields of the `package.json` file.
```
Comment (mapachurro) [inline docs/guide/snaps-development-guide.md:136]: I don't think we have an explanation of how to do this anywhere; obviously it's outside the scope of Snaps itself, but for example should we link to https://docs.npmjs.com/creating-and-publishing-scoped-public-packages ?

File: docs/guide/snaps-development-guide.md [hunk 4]
```diff
@@ -1,47 +1,290 @@
...
+
+To give you an idea of a common error and how to fix it, "sloppily" declared variables (i.e. assigning to a new variable without an explicit variable declaration) are forbidden in strict mode, and therefore in SES as well.
+If you get an error during the `eval` step that says something like `variableName is not defined`, simply prepending `var variableName;` to your snap bundle may solve the problem.
+(This actually happened so frequently with [Babel's](https://babeljs.io/) `regeneratorRuntime` that `mm-snap build` automatically handles that one.)
+
+::: warning Did you modify the snap bundle after running `mm-snap build`?
+Don't forget to run `mm-snap manifest --fix` if you modified your snap bundle after building.
+:::
+
+If you run into a build or eval issue that you can't solve on your own, please create an issue on the [MetaMask/snaps-skunkworks](https://github.com/MetaMask/snaps-skunkworks) repository.
```
Comment (mapachurro) [inline docs/guide/snaps-development-guide.md:265]: Skunkworks repo? Are we deprecating that?

**Review by mapachurro — CHANGES_REQUESTED** [review]
Left a few comments, just the one typo is an actual 'blocker', the rest are just suggestions. Chapeau, Mr. Marks!

File: docs/guide/snaps-development-guide.md (hunk 3 above)
Comment (rekmarks) [reply docs/guide/snaps-development-guide.md:136]: I'll add a link to: https://docs.npmjs.com/packages-and-modules/contributing-packages-to-the-registry

File: docs/guide/snaps-development-guide.md (hunk 4 above)
Comment (rekmarks) [reply docs/guide/snaps-development-guide.md:265]: Not any time soon! Some packages graduate out of there, but others stick around.

**Comment by rekmarks** [comment]
I am admin-merging this PR because we need to publish the updated development guide urgently.

**Comment by rekmarks** [comment]
The inaccuracies reported by @ritave will be addressed elsewhere (possibly by modifying existing behavior).

**Pull request merged**